Map2Mesh Thin Tiles & Nice Outputs

timoffex opened the pull request:
The tile-meshing system is better organized and can handle tiles with varying thicknesses and positions. It can also handle tiles with different top and side images, so this will be easy to enable later on.

The output mesh does not have redundancies in the sense that no faces overlap and vertices are largely reused.

Main changes to look over:
- Map2Mesh class is a little cleaner.
- M2MTileMesher was changed to M2M::TileMesher (there is an M2M namespace with a TileMesher class)
- The TileMesher class has a static getMesherForTile() function that does two things: 1) it decides whether the tile’s old mesh is still good and 2) it returns the new mesher that should be used (if any) for the tile.
- The TileMesher::makeMesh() method returns a vector of SimpleTexturedObjects to support multiple textures. This is implemented quite cleanly by TileBlockyMesher currently.

Everything works fine on my system, but let me know if there are any weird bugs or crashes (besides the crash-on-exit that happens in MeshView).

timoffex committed:
Modified AbstractRenderer::checkGLErrors() to return bool

timoffex committed:
Fixed a bug in SimpleTexturedRenderer

Using callFunctionOnOpenGL() with removeImageTexture() caused data to
get removed sometimes (because removeImageTexture() would be called too
late).

timoffex committed:
Removed SimpleTexturedRenderer::removeImageTexture

timoffex committed:
M2M::PartialMeshData

The PartialMeshData class is used in the new TileMesher system to
construct a mesh piece-by-piece.

It allows one to construct meshes with several textures by
automatically sorting faces into different objects by texture.

Currently, this outputs a list of SimpleTexturedObjects.

timoffex committed:
Made M2M namespace, and remade TileMesher system

At the time of this commit, the program should compile and work
correctly, using the updated meshing system.

Instead of an “M2MTileMesher” class, there is now an M2M namespace with
a TileMesher class.

A TileMesher does two things:

1) Keeps information about the tile neighborhood it was created with
for the sake of being able to check whether the tile’s mesh may need an
update.

2) Has a makeMesh() method that returns a list of meshes that represent
the tile. This is an abstract method implemented by only one subclass
currently, but I will be adding a second subclass (TileDiagonalMesher)
shortly.

timoffex committed:
Merge branch 'master' into map2mesh-system-update

Also changed SimpleTexturedObject and TileMesher systems to use
SharedImageAndSource instead of QSharedPointer<Image>

timoffex committed:
Made thickness editable

Modified some texture-coordinate generation in the TileMesher.

Added new default ground texture.

timoffex committed:
Made MeshView UI a little better.

timoffex committed:
Major cleanup to whole M2M system

Created Array2D_Private namespace to hide away some Array2D features.
Eventually, the iterator classes below Array2D in array2d.h should be
moved here (they are not currently).

Created array2dtools.h header with some functions commonly used with
grid operations.

Added Quad::makeVerticalQuad() to help make vertical quads.

Added M2M_Private::SideTools class with some nice helper functions for
computing side meshes for tiles.

Reduced and cleared up M2M::TileMesher::makeVerticalSideMesh() code.

Made Map2Mesh use an update queue.

timoffex committed:
Merge branch 'master' into map2mesh-system-update

timoffex committed:
added include to fix compile error

Bdtrotte commented on WallsAndHoles/WallsAndHoles.pro:
```diff
@@ -144,11 +146,14 @@ HEADERS += \
 FORMS += \
-    meshview.ui \
-    meshviewcontainer.ui
+    meshview.ui
```
Is this still needed?

timoffex replied:
Nope, I will get rid of that.

Bdtrotte commented on WallsAndHoles/array2d.h:
```diff
@@ -76,6 +88,22 @@ class Array2D {
+     * @return          An object with begin() and end() methods that return iterators
+     *                  that loop over valid (x,y) pairs in this array, represented as QPoints.
+     */
+    IndexCollection indices() const { return IndexCollection(*this); }
```
Is this just as apposed to for(int x ...) { for(int y ...) { ... } }?

If so, I'm not really sure we need it. Same for indexedData(). Why not just have the base array2D iterator include the position when dereferenced?

timoffex replied:
It's just nice. No reason not to have it. Saves 1 indentation and 1-2 lines of code. Also makes intentions quite clear.

Bdtrotte commented on WallsAndHoles/array2d.h:
```diff
@@ -76,6 +88,22 @@ class Array2D {
+     * @return              An object with begin() and end() methods that return iterators that,
+     *                      when dereferenced, return QPair<int, const Type &>.
+     */
+    IndexedConstDataCollection indexedData() const { return IndexedConstDataCollection(*this); }
```
See above

timoffex replied:
I think I didn't end up using this one in my current code, but it's still nice to have (and it just makes sense to have it with the indices() method).

Bdtrotte commented on WallsAndHoles/m2mpartialmesh.h:
```diff
@@ -0,0 +1,182 @@
+
+
+/// Class to wrap image info. This will be changed to use SharedImageAndSource when possible.
+struct ImageInfo {
```
Why not just use SharedImageAndSource?

timoffex replied:
I wrote this before SharedImageAndSource existed. I'm keeping the ImageInfo class to:

1) limit coupling to the definition of ImageInfo
2) allow for image helper methods in the future

In general, it makes TileMesher code easier to change. If I want different image information to be passed around, I can do that by changing ImageInfo. I will actually change this into a class and make image() an accessor method (to reduce coupling even further!).

Bdtrotte commented on WallsAndHoles/m2mpartialmesh.h:
```diff
@@ -0,0 +1,182 @@
+     *                          when looking at the quad. Otherwise, the vertices are output in the order
+     *                              2 1
+     *                              4 3
+     *                          Texture coordinates are always (0,0), (1,0), (1,1), (0,1).
```
Shouldn't this be scaled with height and thickness?

timoffex replied:
Yes! Good point.

Bdtrotte commented on WallsAndHoles/m2mtilemesher.h:
```diff
@@ -2,66 +2,199 @@
+    bool operator ==(const TileNeighborhoodInfo &other) const;
+    bool operator !=(const TileNeighborhoodInfo &other) const;
+
+    const TileInfo &operator ()(int x, int y) const;
```
I don't really like this syntax, I would prefer ::at(x, y); or even [x, y].

::(x, y) seems unconventional.

timoffex replied:
[x, y] is impossible in C++. The (x, y) syntax is used in the C++ Eigen library, which is a very popular linear algebra library (used in Tensorflow!).

timoffex replied:
I'm not sure what you mean by the double colon, but the usage is simply

nbhd(x, y)

Bdtrotte replied:
The double colon was just meant to represent <class>::(x, y). 

Im still not a huge fan, as convention in qt seems to be ::at(x, y). So non uniformity could lead to confusion, and therefore a viscous circle of architectural technical debt!!

Bdtrotte commented on WallsAndHoles/m2mtilemesher.h:
```diff
@@ -2,66 +2,199 @@
  * @brief A collection of methods for generating tile mesh data.
  */
-class M2MTileMesher {
+class TileMesher {
```
TileMesher -> AbstractTileMesher

Bdtrotte commented on WallsAndHoles/m2mtilemesher.h:
```diff
@@ -2,66 +2,199 @@
+     * @param sideDirection         Which side should be made.
+     * @param offset                The offset of the tile.
+     */
+    void makeVerticalSideMesh(PartialMeshData &meshData,
```
It might be nicer if instead of taking a PartialMeshData and modifying it, this returned a PartialMeshData, and could be merged with another PartialMeshData, via a + and += operator.
```
PartialMeshData mySickMesh;

mySickMesh += makeVerticalSideMesh(...);

return mySickMesh
```

timoffex replied:
The way I made PartialMeshData, it is most natural to modify it.

Bdtrotte replied:
I was just saying it seems kind of C style to pass the object along into the function to be modified, where an operator like += or << may be more natural.

Bdtrotte commented on WallsAndHoles/m2mtilemesher.cpp:
```diff
@@ -1,233 +1,360 @@
-        QVector3D(0, 1, 0)
-    };
 
+QVector<QSharedPointer<SimpleTexturedObject>> M2M::TileBlockyMesher::makeMesh(QVector2D offset)
```
Please have all TileMesher:: definitions together

Bdtrotte commented on WallsAndHoles/m2mtilemesher.cpp:
```diff
@@ -1,233 +1,360 @@
-            };
+/* BEGIN TileInfo */
 
+M2M::TileInfo::TileInfo()
```
Please mimic the order used in the .h with the definitions

timoffex replied:
That order will likely change.

Bdtrotte commented on WallsAndHoles/meshviewcontainer.ui:
```diff
@@ -1,65 +0,0 @@
-<?xml version="1.0" encoding="UTF-8"?>
```
Good riddance

Bdtrotte commented on WallsAndHoles/tiletemplatepropertymanager.cpp:
```diff
@@ -44,7 +44,7 @@ QVector<QVector<QVariant>> TileTemplatePropertyManager::properties()
         {"Height",     mTileTemplate->height(),       true,  -1000, 1000},
-        {"Thickness",  mTileTemplate->thickness(),    false,   0.1,    1},
+        {"Thickness",  mTileTemplate->thickness(),    true,    0.1,    1},
         {"X Position", mTileTemplate->position().x(), false, -0.49, 0.49},
```
Can't position also be allowed to be modified now?

timoffex replied:
Yep.

Bdtrotte reviewed (changes requested):
Rough first pass over everything. Still need to pull the branch, and read over all the code in more detail, as well as test everything.

timoffex committed:
Merge branch 'master' into map2mesh-system-update

timoffex committed:
Removed meshview.ui and fixed exit crash bug

timoffex committed:
Refactored ImageInfo to be a class

Instead of having a public image member, it now has an image()
accessor. Constructors are also provided.

timoffex committed:
Renamed TileMesher to AbstractTileMesher

timoffex committed:
Enabled and fixed tile template position editing

The ranges were incorrect.

Bdtrotte committed:
added inline prefix to QPoint's qHash in array2dtools.h to allow multiple definitions

timoffex committed:
fixed thin tiles not outputting walls when on the edge of the map
